### DISCUSSION TITLE
T7-streaming with grapher (stream_t7.py)

### DISCUSSION THREAD

**Discussion by magnus-haw** [LaboratoireMecaniqueLille/crappy#35] [Q&A] [answered]
Hi all,

Trying to use the stream_t7.py example but the grapher class (git version) does not seem to work with streaming inputs. 

Q1) Is there a simple modifier function I can implement to convert the "stream" array into the right format for the grapher to use?
Q2) the hdf5 recorder does not appear to save the time sequence, only the "stream" data array. Is there a simple way of also saving the time data?

**Comment by WeisLeDocto** [comment, marked as answer]
Hello @magnus-haw,

1. The Grapher indeed doesn't work directly with "stream" inputs. The normal way to plot data from streams with the Grapher Block is to use the [Demux](https://github.com/LaboratoireMecaniqueLille/crappy/blob/0485bfc77b3196a07fcbe7e335178bd37b38f1e7/crappy/modifier/demux.py) Modifier on the Link to the Grapher. ~~In the [stream_t7.py example](https://github.com/LaboratoireMecaniqueLille/crappy/blob/0485bfc77b3196a07fcbe7e335178bd37b38f1e7/Examples/stream_t7.py), a very basic Modifier that works in a similar way is used instead, but it also does the job.
The issue you get with this example comes from a typo [line 32](https://github.com/LaboratoireMecaniqueLille/crappy/blob/0485bfc77b3196a07fcbe7e335178bd37b38f1e7/Examples/stream_t7.py#L32), the `'t'` should be `'t(s)'` instead. I will correct it now.~~

2. There indeed doesn't seem to be any way of recording both the timestamps and the data with the current implementation. When using "streams" in Crappy, I think that a good regularity of the acquisition board is assumed, so that the time information can be calculated afterwards when post-processing. I have never worked on the core of the Hdf_recorder Block myself, neither with the underlying `tables` module. Maybe @Dad0u can help here.
A possible quick-and-dirty fix would be to have two Hdf_recorder Blocks, with one of them recording the time information (set the label to record using the `label` argument).

Weis

**EDIT**: 
After double-checking, seems like just changing the `'t'` to `'t(s)'` won't fix the problem, as the data from all channels is sent by the T7streamer InOut over a common label (`stream`) instead of one label per channel (`AIN0` and `AIN1`). A [Demux](https://github.com/LaboratoireMecaniqueLille/crappy/blob/0485bfc77b3196a07fcbe7e335178bd37b38f1e7/crappy/modifier/demux.py) Modifier is after all necessary instead of the current custom Modifier. I'll upload a definitive fix tomorrow when I have access to a Labjack.

**Reply by WeisLeDocto** [reply]
Coming back to you after having tried a few things out.

I was able to write a working code for what you want to achieve, using a Labjack T7 Pro. Here's the exact script : 
```
# coding: utf-8

"""
Demonstrates how to use the stream mode on Labjack T7 boards.

Hardware required:
  - Labjack T7
"""

import crappy
import tables


if __name__ == "__main__":
  s = crappy.blocks.IOBlock("T7_streamer",
                            channels=[{'name': 'AIN0', 'gain': 2,
                                       'offset': -13},
                                      {'name': 'AIN1', 'gain': 2,
                                       "make_zero": True}],
                            streamer=True)

  g = crappy.blocks.Grapher(('t(s)', 'AIN0'), ('t(s)', 'AIN1'))
  crappy.link(s, g, modifier=crappy.modifier.Demux(labels=('AIN0', 'AIN1')))

  rec = crappy.blocks.Hdf_recorder("out.h5", atom=tables.Float64Atom())
  rec_t = crappy.blocks.Hdf_recorder("out_t.h5", atom=tables.Float64Atom(),
                                     label='t(s)')
  crappy.link(s, rec)
  crappy.link(s, rec_t)
  crappy.start()

```

In order for it to work, I had to modify a few lines in the module. That was a nice occasion to publish a new 1.5.10 release. The provided script should run with the latest version of pip install, or with the master branch of the repo. Not super sure about the stability of this new release, but I'm now focusing on the 2.0.0 release to come.

Please tell me if you would still have issues with the "streams" or with Crappy in general. And if you would have suggestions of enhancements for the 2.0.0 release, feel free to share them.

Weis

**Reply by magnus-haw** [reply]
Thanks! I'll probably take a stab at modifying the Hdf_recorder block to also save the time data so we'll see how that goes.

The one mod I would suggest is adding the following line into the grapher.py block: **plt.tight_layout()** (add @line 144, right before plt.show(block=False) ). This cleans up the layout for arbitrary aspect ratio plots where some of the labels can be obscured.
